Lib/concurrent/futures/_base.py

```diff
@@ -544,18 +545,40 @@ def map(self, fn, *iterables, timeout=None, chunksize=1):
         """
         if timeout is not None:
             end_time = timeout + time.time()
+        if prefetch is None:
+            prefetch = self._max_workers
+        if prefetch < 0:
+            raise ValueError("prefetch count may not be negative")
 
-        fs = [self.submit(fn, *args) for args in zip(*iterables)]
+        argsiter = zip(*iterables)
+
+        fs = collections.deque(self.submit(fn, *args) for args in itertools.islice(argsiter, self._max_workers + prefetch))
 
         # Yield must be hidden in closure so that the futures are submitted
         # before the first iterator value is required.
         def result_iterator():
```

Comment:
My mistake, please ignore the above comment. Both the original code and this PR need this closure. Without it, the entire map function would become a generator, and so it won't start executing until someone starts iterating through it. So with the code like this: `it = executor.map(fn)` , nothing at all will happen until someone starts iterating through `it`. With the closure, this code would initiate the acquisition of inputs and their submission to the worker.